src/documentation/programming-guide.md

```diff
@@ -532,37 +533,176 @@ tree, [2]
+      emailTag -> [],
+      phoneTag -> ["222-333-4444"]
+   }
+   "julia" {
+      emailTag -> ["julia@example.com"],
+      phoneTag -> []
+   }
 ```
+```py
+{% github_sample /apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets_test.py tag:model_group_by_key_cogroupbykey_tuple_outputs
+%}```
+
+##### **Applying CoGroupByKey**
 
-`CoGroupByKey` gathers up the values with the same key from all `PCollection`s, and outputs a new pair consisting of the unique key and an object `CoGbkResult` containing all values that were associated with that key. If you apply `CoGroupByKey` to the input collections above, the output collection would look like this:
+The following example joins the two `PCollection`s with `CoGroupByKey`, followed by
+a `ParDo` to consume the result. Then, the code uses tags to look up data from
+each collection.
+
+```java
+  // Each set of key-value pairs is read into separate PCollections.
+  // Each shares a common key ("K").
+  PCollection<KV<K, V1>> pt1 = ...;
+  PCollection<KV<K, V2>> pt2 = ...;
+
+  // Create tuple tags for the value types in each collection.
+  final TupleTag<V1> t1 = new TupleTag<V1>();
+  final TupleTag<V2> t2 = new TupleTag<V2>();
+
+  // Merge collection values into a CoGbkResult collection
+  PCollection<KV<K, CoGbkResult>> coGbkResultCollection =
+    KeyedPCollectionTuple.of(t1, pt1)
+                         .and(t2, pt2)
+                         .apply(CoGroupByKey.<K>create());
+
+  // Access results and do something with them.
+  PCollection<T> finalResultCollection =
+    coGbkResultCollection.apply(ParDo.of(
+      new DoFn<KV<K, CoGbkResult>, T>() {
+        @Override
+        public void processElement(ProcessContext c) {
+          KV<K, CoGbkResult> e = c.element();
+          // Get all collection 1 values
+          Iterable<V1> pt1Vals = e.getValue().getAll(t1);
+          // Now get collection 2 values
+          V2 pt2Val = e.getValue().getOnly(t2);
+          // ... Do something ...
+          c.output(...some T...);
+        }
+      }));
 ```
-user1, [[address1], [order1, order2]]
-user2, [[address2], [order3]]
-user3, [[address3], []]
-guest, [[], [order4]]
-...
-````
+```py
+{% github_sample /apache/beam/blob/master/sdks/python/apache_beam/examples/snippets/snippets.py tag:model_group_by_key_cogroupbykey_tuple
+%}```
+
+##### **CoGroupByKey with unbounded PCollections**
+
+> **NOTE:** This section applies only to the Beam SDK for Java. The Beam SDK
+for Python does not yet support this feature.
+
+You must use either [non-global windowing](#setwindowingfunction) or an
+[aggregation trigger](#triggers) in order to perform a `CoGroupByKey` with
+unbounded `PCollection`s. This is because a bounded `CoGroupByKey` must wait
+for all the data with a certain key to be collected, but with unbounded
+collections, the data is unlimited. Windowing and/or triggers allow grouping to
+operate on logical, finite bundles of data within the unbounded data streams.
```

Comment:
This is a reiteration of the `GroupByKey` requirements, yes? Could those be condensed?

Reply:
I moved this up to an "Unbounded PCollections" blurb in GroupByKey, and left a minimal pointer from CoGroupByKey to the GroupByKey section for more details. see what you think.